source/common/tcp_proxy/upstream.cc

```diff
@@ -152,5 +154,99 @@ void HttpUpstream::doneWriting() {
   }
 }
 
+TcpConnPool::TcpConnPool(const std::string& cluster_name, Upstream::ClusterManager& cluster_manager,
+                         Upstream::LoadBalancerContext* context,
+                         Tcp::ConnectionPool::UpstreamCallbacks& upstream_callbacks)
+    : upstream_callbacks_(upstream_callbacks) {
+  conn_pool_ = cluster_manager.tcpConnPoolForCluster(cluster_name,
+                                                     Upstream::ResourcePriority::Default, context);
+}
+
+TcpConnPool::~TcpConnPool() {
+  if (upstream_handle_ != nullptr) {
+    upstream_handle_->cancel(Tcp::ConnectionPool::CancelPolicy::CloseExcess);
+  }
+}
+
+bool TcpConnPool::valid() const { return conn_pool_ != nullptr; }
+
+void TcpConnPool::newStream(GenericConnectionPoolCallbacks* callbacks) {
+  callbacks_ = callbacks;
+  // Given this function is reentrant, make sure we only reset the upstream_handle_ if given a
+  // valid connection handle. If newConnection fails inline it may result in attempting to
+  // select a new host, and a recursive call to initializeUpstreamConnection. In this case the
+  // first call to newConnection will return null and the inner call will persist.
+  Tcp::ConnectionPool::Cancellable* handle = conn_pool_->newConnection(*this);
+  if (handle) {
+    ASSERT(upstream_handle_ == nullptr);
+    upstream_handle_ = handle;
+  }
+}
+
+void TcpConnPool::onPoolFailure(ConnectionPool::PoolFailureReason reason,
+                                Upstream::HostDescriptionConstSharedPtr host) {
+  upstream_handle_ = nullptr;
+  callbacks_->onGenericPoolFailure(reason, host);
+}
+
+void TcpConnPool::onPoolReady(Tcp::ConnectionPool::ConnectionDataPtr&& conn_data,
+                              Upstream::HostDescriptionConstSharedPtr host) {
+  upstream_handle_ = nullptr;
+  Tcp::ConnectionPool::ConnectionData* latched_data = conn_data.get();
+  Network::Connection& connection = conn_data->connection();
+
+  auto upstream = std::make_unique<TcpUpstream>(std::move(conn_data), upstream_callbacks_);
+  callbacks_->onGenericPoolReady(&connection.streamInfo(), std::move(upstream), host,
+                                 latched_data->connection().localAddress(),
+                                 latched_data->connection().streamInfo().downstreamSslConnection());
+}
+
+HttpConnPool::HttpConnPool(const std::string& cluster_name,
+                           Upstream::ClusterManager& cluster_manager,
+                           Upstream::LoadBalancerContext* context, std::string hostname,
+                           Tcp::ConnectionPool::UpstreamCallbacks& upstream_callbacks)
+    : hostname_(hostname), upstream_callbacks_(upstream_callbacks) {
+  conn_pool_ = cluster_manager.httpConnPoolForCluster(
+      cluster_name, Upstream::ResourcePriority::Default, absl::nullopt, context);
+}
+
+HttpConnPool::~HttpConnPool() {
+  if (upstream_handle_ != nullptr) {
+    upstream_handle_->cancel(Tcp::ConnectionPool::CancelPolicy::Default);
```

Comment:
This cancel uses `Default`, but the Tcp one above uses `CloseExcess`. Why are they different?

Reply:
This would read better if you used `Http::ConnectionPool::...`. They're the same type, but it looks weird this way.

Reply:
legacy - the TCP pool always used close excess, and HTTP always used default.

Reply:
Add comment explaining that please.